feat(har): record sizes, compression and httpVersion

tnolet opened the pull request:
This PR attempts to update a set of missing properties in the HAR file, where earlier a default -1 was reported and also report the correct HTTP Version, which was hard coded to `HTTP/1.1`

The test in `tests/har.spec.ts:244` should show what is being done.
- `bodySize`
- `headersSize`
- `httpVersion`
- `_transferSize`: note this is not an official property, but used by Chrome.
- `content.size`
- `content.compression`

The biggest change is that the fields at the end of the request/response cycle are now done on `Page.Events.onRequestFinished` instead of `Page.Events.onResponse`.


## Note 1
This does not work yet for Firefox as no `transferSize` like property is available. However, we can probably make that happen with some work on the NetworkObserver.js patches.

## Note 2
I went down the rabbit hole of solving this problem by populating the various `size` using the `performance.getEntriesByName()` API from the browser. However, CORS settings makes this unreliable.

tnolet committed:
feat(har): wip: adds response content lengths to har file

tnolet committed:
feat(har): surface protocol / http version information

The http version was hard coded to http/1.1 which was wrong. Als h2 and h3 have different ways of
reporting headers and lenghts

tnolet committed:
style(har): missing semicolon

tnolet committed:
feat(har): resolves httpVersion

HTTP version is now resolved in one place or substituted by a fallback. Works with reporting h2 and
later.

tnolet committed:
refactor(har): use httpVersion variable for calculating header size

tnolet committed:
feat(har): don't set application/octet-stream as fallback mimetype

tnolet committed:
feat(har): only set content.text and content.encoding when buffer has value

tnolet committed:
feat(wk): adds setEncodedDataLength for wk response

tnolet committed:
fix(har): set http version to lowercase again as per current test

tnolet committed:
feat(har): also do not set application/octet-stream as default mimetype from response header

dgozman commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -68,7 +70,7 @@ export class HarTracer {
     this._log.pages.push(pageEntry);
     page.on(Page.Events.Request, (request: network.Request) => this._onRequest(page, request));
-    page.on(Page.Events.Response, (response: network.Response) => this._onResponse(page, response));
+    page.on(Page.Events.RequestFinished, (request: network.Request) => this._onRequestFinished(page, request));
```
Not all requests finish, some of them stall, and others fail. I think we should update half of the fields on Response event, and another half on RequestFinished/RequestFailed. WDYT?

tnolet replied:
yes, this is a correct observation. Somehow it is feels messy though. Reading the code it's not super clear what gets done when. I do think we have no other way, this is how HTTP / Browser networking is.

Testing this is tricky. We would need a local server for tests that does not respond or stalls some requests.

dgozman replied:
I think you can just add a test with a failed request and see that some information is there. I do not think that extra testing complexity will be worth it in this particular case.

dgozman commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -152,38 +155,50 @@ export class HarTracer {
+
+    const response = await request.response();
+
+    if (!response) return;
```
style: we usually return on a separate line.

tnolet replied:
gotcha

dgozman commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -152,38 +155,50 @@ export class HarTracer {
+
+    if (!response) return;
+
+    const httpVersion = response.protocol() ? response.protocol() : FALLBACK_HTTP_VERSION;
```
There is already a fallback in the `protocol()` getter. Let's have it in one place (I prefer this one).

tnolet replied:
@dgozman do you mean: "switch to my `FALLBACK_HTTP_VERSION`" or do you mean "keep current fallback"?

dgozman replied:
I mean "make `response.protocol()` nullable" and keep `FALLBACK_HTTP_VERSION` here.

tnolet replied:
aha, e.g.

```
 // Response class
  protocol(): string | undefined {
    return this._protocol;
  }
```

tnolet replied:
ok, this was an easy fix. Pushed it already

dgozman commented on src/server/webkit/wkPage.ts:
```diff
@@ -966,8 +966,11 @@ export class WKPage implements PageDelegate {
     // event from protocol. @see https://crbug.com/883475
     const response = request.request._existingResponse();
-    if (response)
+    if (response) {
```
Do we have `protocol` available in WebKit?

tnolet replied:
Good catch, we need to add that at `src/server/webkit/wkInterceptableRequest.ts:113`

tnolet replied:
The issue is it only becomes available later in the cycle. The `metrics` property (which holds the `protocol` gets populated in the `loadingFinishedPayload` at `src/server/webkit/protocol.ts:5475`

tnolet replied:
@dgozman this was actually not that hard to fix. I added a patch for it in the last commit. We are now using the http protocol reported by webkit.

dgozman commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -152,38 +155,50 @@ export class HarTracer {
+
+    const httpVersion = response.protocol() ? response.protocol() : FALLBACK_HTTP_VERSION;
+    const transferSize = response.encodedDataLength();
+    const headersSize = calcResponseHeadersSize(httpVersion, response.status(), response.statusText(), response.headers());
```
Note that in Chromium we have `headersText` arriving in `Network.responseReceivedExtraInfo`, so that you don't have to calculate it. Let's add a TODO to use that.

tnolet replied:
@dgozman I think a TODO would be fine, right now my solution is useful as it is uniform for all supported browsers.

dgozman commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -120,27 +122,28 @@ export class HarTracer {
-        postData: undefined,
+        postData: postDataForHar(request) || undefined,
         headersSize: -1,
         bodySize: -1,
```
Do you want to set `bodySize` here as well? See how DevTools does it here: https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/models/har/Log.ts;l=378-388?q=f:front_end%20-f:out%2F%20responseHeadersText.

dgozman reviewed (commented):
Looks good overall, left a few comments.

tnolet committed:
style(har): small code style fix

tnolet committed:
refactor(har): simplify default/fallback http version getter

tnolet committed:
feat(response): adds setProtocol() method to add protocol after constructing Response

webkit reports protocol in the .metrics object later than chromium, so protocol is not available
during constructing the response. This method allows setting it later in the cycle.

tnolet committed:
chore(har): adds TODO for replacing headersSize with "native" method

tnolet committed:
refactor(har): splits populating har over Response and RequestFinished events

aslushnikov commented:
Hey @tnolet! What's the status of this? 

For the firefox, it'll be very easy to export transfer size: https://ffsearch.azurewebsites.net/#path=%2Fhome%2Fjoe%2Fplaywright%2Fbrowser_patches%2Ffirefox%2Fcheckout%2Fnetwerk%2Fprotocol%2Fhttp%2FnsIHttpChannel.idl&line=90

I can help with this if needed!

tnolet commented:
> Hey @tnolet! What's the status of this?
> 
> For the firefox, it'll be very easy to export transfer size: https://ffsearch.azurewebsites.net/#path=%2Fhome%2Fjoe%2Fplaywright%2Fbrowser_patches%2Ffirefox%2Fcheckout%2Fnetwerk%2Fprotocol%2Fhttp%2FnsIHttpChannel.idl&line=90
> 
> I can help with this if needed!

@aslushnikov I think we are pretty damn far with this. It would be great if you can help patch FF so we can keep this PR simple and not make any exemptions for FF. How should we coordinate this? I see the following steps:

- have FF export transferSizes
- integrate the FF tranferSizes so they are available to the harTracer.
- fix any merge conflicts
- profit?

mxschmitt commented:
Closing since https://github.com/microsoft/playwright/pull/7470 is merged.
